src/plugins/workspace/server/permission_control/client.ts

```diff
@@ -0,0 +1,187 @@
+      if (!savedObject.permissions) {
+        return true;
+      }
+
+      const aclInstance = new ACL(savedObject.permissions);
+      const hasPermission = aclInstance.hasPermission(permissionModes, principals);
+      if (!hasPermission) {
+        notPermittedSavedObjects.push(savedObject);
+      }
+      return hasPermission;
+    });
+    if (!hasAllPermission) {
+      this.logNotPermitted(notPermittedSavedObjects, principals, permissionModes);
+    }
+    return hasAllPermission;
+  }
+
+  /**
+   * In batch validate case, the logic is a.withPermission && b.withPermission
+   * @param request
+   * @param savedObjects
+   * @param permissionModes
+   * @returns
+   */
+  public async batchValidate(
+    request: OpenSearchDashboardsRequest,
+    savedObjects: SavedObjectsBulkGetObject[],
+    permissionModes: SavedObjectsPermissionModes
+  ) {
+    const savedObjectsGet = await this.bulkGetSavedObjects(request, savedObjects);
+    if (!savedObjectsGet.length) {
+      return {
+        success: false,
+        error: i18n.translate('savedObjects.permission.notFound', {
+          defaultMessage: 'Can not find target saved objects.',
+        }),
+      };
+    }
+
+    if (savedObjectsGet.some((item) => item.error)) {
+      return {
+        success: false,
+        error: savedObjectsGet
+          .filter((item) => item.error)
+          .map((item) => item.error?.error)
+          .join('\n'),
+      };
+    }
+
+    const principals = getPrincipalsFromRequest(request);
+    let savedObjectsBasicInfo: any[] = [];
+    const hasAllPermission = savedObjectsGet.every((item) => {
+      // for object that doesn't contain ACL like config, return true
+      if (!item.permissions) {
+        return true;
+      }
+      const aclInstance = new ACL(item.permissions);
+      const hasPermission = aclInstance.hasPermission(permissionModes, principals);
+      if (!hasPermission) {
+        savedObjectsBasicInfo = [
+          ...savedObjectsBasicInfo,
+          {
+            id: item.id,
+            type: item.type,
+            workspaces: item.workspaces,
+            permissions: item.permissions,
+          },
+        ];
+      }
+      return hasPermission;
+    });
+    if (!hasAllPermission) {
+      this.logger.debug(
+        `Authorization failed, principals: ${JSON.stringify(
+          principals
+        )} has no [${permissionModes}] permissions on the requested saved object: ${JSON.stringify(
+          savedObjectsBasicInfo
+        )}`
+      );
+    }
+    return {
+      success: true,
+      result: hasAllPermission,
+    };
+  }
+
+  public async getPrincipalsOfObjects(
+    request: OpenSearchDashboardsRequest,
+    savedObjects: SavedObjectsBulkGetObject[]
+  ): Promise<Record<string, TransformedPermission>> {
+    const detailedSavedObjects = await this.bulkGetSavedObjects(request, savedObjects);
+    return detailedSavedObjects.reduce((total, current) => {
```

Comment:
This function is for get the principals of specific object.  It's a part of permission validation. Since we do the permission validation in the server side. We need to get the principals of specific object without check user's permission. This function should not expose to the frontend. This function was only be used in `${WORKSPACES_API_BASE_URL}/principals`, I think we can remove it.

Reply:
I see, if there is no clear use case, I suggest to remove it.